lib/srv/db/mysql/engine.go

```diff
@@ -164,9 +164,16 @@ func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Sessio
 func (e *Engine) updateServerVersion(sessionCtx *common.Session, serverConn *client.Conn) error {
 	serverVersion := serverConn.GetServerVersion()
 	statusVersion := sessionCtx.Database.GetMySQLServerVersion()
-	// Update only when needed
+	// Update only when needed.
+	// Note that sessionCtx.Database is a copy of the database cached by
+	// database service. Call UpdateProxiedDatabase to do the update instead of
+	// setting the copy.
 	if serverVersion != "" && serverVersion != statusVersion {
-		sessionCtx.Database.SetMySQLServerVersion(serverVersion)
+		return trace.Wrap(e.UpdateProxiedDatabase(sessionCtx.Database.GetName(), func(db types.Database) error {
```

Comment:
I'm not a fan of this level of nesting - the happy path is rather hidden. I find Go code to be quite "scannable" if it does not pack too much into one line / function call. Perhaps this?
```
err := e.UpdateProxiedDatabase(sessionCtx.Database.GetName(), func(db types.Database) error {
    db.SetMySQLServerVersion(serverVersion)
    e.Log.WithField("server-version", serverVersion).Debug("Updated MySQL server version.")
    return nil
})
if err != nil {
    return trace.Wrap(err)
}
```

and let the happy path fall through to the end of the function? This also helps if something needs to be added after this block later - the inside of the block does not need to be altered as the happy path continues after it.

Reply:
refactored the function a little bit to make it more readable.